File: gen_dynamic_network/processTimeFrameData.py
# Author: Anthony Kai Kwang Ma
# Email: akma327@stanford.edu
# Date: February 18, 2016

# Process Noncovalent Interaction by Time Frame for MOR Trajectories- Anthony Ma 02/18/16

from __future__ import print_function
import os
import sys
import datetime
import time
import itertools
import copy
import mdtraj as md
from hbond import baker_hubbard
from saltbridge import *
from vanderwaal import *
from hydrophobe import *
from pication import *
from aromatic import *
from waterMediatedHBonds import * 

# Usage:
# python processTimeFrameData.py <Stitch Interaction Input Path> <Input File Name> <Binary Dictionary Output Path> <Output File Name> <interaction type><binarize flag>
# Example:
# StitchedInteractionPath="/scratch/PI/rondror/akma327/noncovalent_Interaction_Scripts/mOR-InteractionOutput/MOR_active_waters/mor_active_refine150_agonist_noNanobody_noNTerm/rep_2"
# INPUT_FILE_PATH="salt_bridge_result_stitch.txt"
# OUTPUT_PATH="/scratch/PI/rondror/akma327/noncovalent_Interaction_Scripts/mOR-InteractionOutputDictionary/MOR_active_waters/mor_active_refine150_agonist_noNanobody_noNTerm/rep_2"
# OUTPUT_FILE_NAME="salt_bridge_result_dict.txt"
# python processTimeFrameData.py $StitchedInteractionPath $INPUT_FILE_PATH $OUTPUT_PATH $OUTPUT_FILE_NAME -sb -b


FLAG_INTERACTION = {"-sb": "Salt Bridges", "-pc": "Pi-Cation", 
"-ps": "Pi-Stacking", "-ts": "T_Stacking", "-vdw": "Van Der Waals", 
"-hb": "Hydrogen Bonds", "-hbw": "Hydrogen Bond-Water Mediated",
"-hbbb": "Backbone Backbone Hydrogen Bonds",
"-hbsb": "Sidechain Backbond Hydrogen Bonds",
"-hbss": "Sidechain Sidechain Hydrogen Bonds",
"-rw": "Residue Water Hydrogen Bonds",
"-wb": "Water Bonds",
"-wb2": "Extended Water Bonds",
"-hlb": "Ligand Backbone Hydrogen Bonds",
"-hls": "Ligand Sidechain Hydrogen Bonds",
"-lw": "Ligand Water Hydrogen Bonds",
"-lwb": "Ligand Water Bonds",
"-lwb2": "Ligand Extended Water Bonds"}

def retrieveDataFileReader():
	print("retrieveDataFileReader()")
	input_path = sys.argv[1]
	input_filename = sys.argv[2]
	filename = input_path + "/" + input_filename
	print("Creating Binary Dictionary for: " + filename)
	if os.path.isfile(filename):
		f = open(filename, 'r')
		return f
	else:
		print("Interaction Results Not Found")
		return None

def createTimeFrameDict():
	numLinesRead = 0
	print("createTimeFrameDict()")
	f = retrieveDataFileReader()
	stride = 1
	TrajectoryPaths = []
	TopologyPath = ""
	maxDelta = 0
	appendcount = 0
	#dictionary to store pair of interaction key with list of time points
	timelapse_interaction_by_chain_dict = {}
	if(f != None):
		currFrame = 0
		currProtein = ""
		for line in f:
			# if(numLinesRead % 1000 == 0): 
			# 	print("Line: " + str(numLinesRead) + " ** " + str(line))
			# 	if(0 in timelapse_interaction_by_chain_dict):
			# 		print("len dict " + str(len(timelapse_interaction_by_chain_dict[0])))
			numLinesRead += 1
			if("Stride:" in line):
				stride = int(line.split("Stride:")[1])
			elif("TrajectoryPath:" in line):
				TrajectoryPaths.append(line.split("TrajectoryPath:")[1])
			elif("TopologyPath:" in line):
				TopologyPath = line.split("TopologyPath:")[1]
			elif("nFrames:" in line):
				continue
			elif("Computing Time:" in line):
				continue
			elif("Frame: " in line):
				currFrame = int(line.split("Frame: ")[1])
			elif("Salt Bridges:" in line):
				currProtein = (line.split("Salt Bridges:")[1])
			elif("Pi-Cation:" in line):
				currProtein = (line.split("Pi-Cation:")[1])
			elif("Pi-Stacking:" in line):
				currProtein = (line.split("Pi-Stacking:")[1])
			elif("T-Stacking:" in line):
				currProtein = (line.split("T-Stacking:")[1])
			elif("Van Der Waals:" in line):
				currProtein = (line.split("Van Der Waals:")[1])
			elif("Hydrogen Bonds:" in line):
				currProtein = (line.split("Hydrogen Bonds:")[1])
			elif("Hydrogen Bond-Water Mediated:" in line):
				currProtein = (line.split("Hydrogen Bond-Water Mediated:")[1])
			elif("Backbone Backbone Hydrogen Bonds:" in line):
				currProtein = (line.split("Backbone Backbone Hydrogen Bonds:")[1])
			elif("Sidechain Backbond Hydrogen Bonds:" in line):
				currProtein = (line.split("Sidechain Backbond Hydrogen Bonds:")[1])
			elif("Sidechain Sidechain Hydrogen Bonds:" in line):
				currProtein = (line.split("Sidechain Sidechain Hydrogen Bonds:")[1])
			elif("Residue Water Hydrogen Bonds:" in line):
				currProtein = (line.split("Residue Water Hydrogen Bonds:")[1])
			elif("Water Bonds:" in line):
				currProtein = (line.split("Water Bonds:")[1])
			elif("Extended Water Bonds:" in line):
				currProtein = (line.split("Extended Water Bonds:")[1])
			elif("Ligand Backbone Hydrogen Bonds:" in line):
				currProtein = (line.split("Ligand Backbone Hydrogen Bonds:")[1])
			elif("Ligand Sidechain Hydrogen Bonds:" in line):
				currProtein = (line.split("Ligand Sidechain Hydrogen Bonds:")[1])
			elif("Ligand Water Hydrogen Bonds:" in line):
				currProtein = (line.split("Ligand Water Hydrogen Bonds:")[1])
			elif("Ligand Water Bonds:" in line):
				currProtein = (line.split("Ligand Water Bonds:")[1])
			elif("Ligand Extended Water Bonds:" in line):
				currProtein = (line.split("Ligand Extended Water Bonds:")[1])
			
			

			elif(line != "\n"):
				# if(numLinesRead >100000):
				# 	break
				key_pair = tuple(line.strip().split(" -- "))
				split_key = key_pair[0].split("_")
				if(len(split_key) > 1):
					chain_id = int(split_key[1])
				else:
					chain_id = 0
				if(chain_id not in timelapse_interaction_by_chain_dict.keys()):
					timelapse_interaction_by_chain_dict[chain_id] = {} #create empty dictionary
					timelapse_interaction_by_chain_dict[chain_id][key_pair] = [currFrame]
				else:
					tic1 = datetime.datetime.now()
					if key_pair not in timelapse_interaction_by_chain_dict[chain_id].keys():
						toc1 = datetime.datetime.now()
						deltaLookup = toc1 - tic1 
						deltaLookup =deltaLookup.microseconds
						appendcount += 1
						tic = datetime.datetime.now()
						timelapse_interaction_by_chain_dict[chain_id][key_pair] = [currFrame]
						toc = datetime.datetime.now()
						delta = toc - tic
						delta = delta.microseconds
						# if (delta > maxDelta): maxDelta = delta
						# if (maxDelta == delta or appendcount %50 == 0): 
						if(numLinesRead %1000 ==0):
							print("Line Number: " + str(numLinesRead) + " Lookup Time1: " + str(deltaLookup))
							print("time for append: " + str(delta) + " Dict Size: " + str(len(timelapse_interaction_by_chain_dict[chain_id])))

					else:
						toc2 = datetime.datetime.now()
						timelapse_interaction_by_chain_dict[chain_id][key_pair].append(currFrame)
						deltaLookup = toc2 -tic1
						deltaLookup =deltaLookup.microseconds
						if(numLinesRead %1000 == 0):
							print("Line Number: " + str(numLinesRead) + " Lookup Time2: " + str(deltaLookup) + " Dict Size: " + str(len(timelapse_interaction_by_chain_dict[chain_id])))
	totalFrames = currFrame + 1
	return stride, TrajectoryPaths, TopologyPath, timelapse_interaction_by_chain_dict, totalFrames

def displayFrequencyMap(stride, TrajectoryPath, TopologyPath, timelapse_interaction_by_chain_dict):
	print("Stride:" + str(stride))
	print("TrajectoryPath:" + TrajectoryPath)
	print("TopologyPath:" + TopologyPath)
	for chain_id in timelapse_interaction_by_chain_dict.keys():
		print("Dictionary for Chain: " + str(chain_id))
		for interaction in timelapse_interaction_by_chain_dict[chain_id].keys():
			key_identifier = interaction[0] + " -- " + interaction[1]
			print(key_identifier + ":\n" + str(timelapse_interaction_by_chain_dict[chain_id][interaction]))

def writeFrequencyMap(stride, TrajectoryPaths, TopologyPath, timelapse_interaction_by_chain_dict, totalFrames):
	print("writeFrequencyMap()")
	output_path = sys.argv[3]
	output_filename = sys.argv[4]
	interaction_selection = sys.argv[5]
	filename = output_path + "/" + output_filename
	if not os.path.exists(os.path.dirname(filename)):
		os.makedirs(os.path.dirname(filename))
	f = open(filename, 'w')
	f.write("Stride:" + str(stride) + "\n")
	for TrajectoryPath in TrajectoryPaths:
		f.write("TrajectoryPath:" + TrajectoryPath + "\n")
	f.write("TopologyPath:" + TopologyPath + "\n")
	f.write(FLAG_INTERACTION[interaction_selection] + " Dictionary Heat Map " + "\n")
	f.write("TotalFrames: " + str(totalFrames) + "\n")
	for chain_id in timelapse_interaction_by_chain_dict.keys():
		f.write("Dictionary for Chain: " + str(chain_id) + "\n")
		for interaction in timelapse_interaction_by_chain_dict[chain_id].keys():
			key_identifier = interaction[0] + " -- " + interaction[1]
			f.write(key_identifier + "~" + str(timelapse_interaction_by_chain_dict[chain_id][interaction]) + "\n")
	print("Wrote results to: " + filename)

# converts a list of time points in which interaction is present to a binary
# representation over all time points. eg. [1,2,5] --> [0,1,1,0,0,1]
def binarizeTimePoints(timelapse_interaction_by_chain_dict, totalFrames):
	print("binarizeTimePoints()")
	def binarizeTime(timePoints, totalFrames):
		binaryVec = [0]*(totalFrames)
		for tp in timePoints:
			binaryVec[tp] = 1
		return binaryVec
	binary_timelapse_interaction_by_chain_dict = copy.deepcopy(timelapse_interaction_by_chain_dict)
	for chain_id in binary_timelapse_interaction_by_chain_dict.keys():
		for interaction in binary_timelapse_interaction_by_chain_dict[chain_id].keys():
			timePoints = binary_timelapse_interaction_by_chain_dict[chain_id][interaction]
			binarizedTime = binarizeTime(timePoints, totalFrames)
			binary_timelapse_interaction_by_chain_dict[chain_id][interaction] = binarizedTime
	return binary_timelapse_interaction_by_chain_dict


# Create the frequency map either in time point form, or binaryized form with -b flag
def calculateFrequencyMap():
	print("calculateFrequencyMap()")
	binarizeFlagIndex = 6
	stride, TrajectoryPaths, TopologyPath, timelapse_interaction_by_chain_dict, totalFrames = createTimeFrameDict()
	if(len(sys.argv) > binarizeFlagIndex and sys.argv[binarizeFlagIndex] == "-b"):
		binary_timelapse_interaction_by_chain_dict = binarizeTimePoints(timelapse_interaction_by_chain_dict, totalFrames)
		writeFrequencyMap(stride, TrajectoryPaths, TopologyPath, binary_timelapse_interaction_by_chain_dict, totalFrames)
	else:
		writeFrequencyMap(stride, TrajectoryPaths, TopologyPath, timelapse_interaction_by_chain_dict, totalFrames)

if __name__ == "__main__":
	tic = time.clock()
	print ("START MOR_processTimeFrameData() TIME: " + datetime.datetime.now().strftime("%A, %d. %B %Y %I:%M%p"))
	calculateFrequencyMap()
	toc = time.clock()
	print ("END MOR_processTimeFrameData() TIME: " + datetime.datetime.now().strftime("%A, %d. %B %Y %I:%M%p"))
	print("Total Time for MOR_processTimeFrameData: " + str(toc-tic))

File: v1/gen_dynamic_network/processTimeFrameData.cpp
// Author: Anthony Kai Kwang Ma
// Email: akma327@stanford.edu
// Date: February 18, 2016

// Process Noncovalent Interaction by Time Frame for MOR Trajectories 
// C++ implementation that will use a binary search tree data structure 
// In the case of water mediated hydrogen bonds, where the dictionary size
// Is very large - Anthony Ma 02/22/16

// Usage:
// ./processTimeFrameData <Stitch Interaction Input Path> <Input File Name> <Binary Dictionary Output Path> <Output File Name> <interaction type><binarize flag>
// Example:
// StitchedInteractionPath="/scratch/PI/rondror/akma327/noncovalent_Interaction_Scripts/mOR-InteractionOutput/MOR_active_waters/mor_active_refine150_agonist_noNanobody_noNTerm/rep_2"
// INPUT_FILE_NAME="salt_bridge_result_stitch.txt"
// OUTPUT_PATH="/scratch/PI/rondror/akma327/noncovalent_Interaction_Scripts/mOR-InteractionOutputDictionary/MOR_active_waters/mor_active_refine150_agonist_noNanobody_noNTerm/rep_2"
// OUTPUT_FILE_NAME="salt_bridge_result_dict.txt"
// ./processTimeFrameData $StitchedInteractionPath $INPUT_FILE_NAME $OUTPUT_PATH $OUTPUT_FILE_NAME -sb -b

#include <stdlib.h> 
#include <iostream>
#include <fstream>
#include <sstream>
#include <string>
#include <string.h>
#include <vector>
#include <map>
#include <time.h>
using namespace std;

std::map<string, string> FLAG_INTERACTION;

int stringToInt(string s){
	int x;
	stringstream convert(s);
	convert >> x;
	return x;
}

// C++ Implementation of python's split implementation that converts a string
// to a vector containing the tokens split upon a desired delimeter
vector<string> split(string str, char delimiter) {
  vector<string> internal;
  stringstream ss(str); // Turn the string into a stream.
  string tok;
  while(getline(ss, tok, delimiter)) {
    internal.push_back(tok);
  }
  return internal;
}


int getChainId(string line){
	string atom1 = split(line, '-')[1];
	if(atom1.find("_") != std::string::npos){
		return stringToInt(split(atom1, '_')[1]);
	}
	return 0;
}

string retrieveDataFileReader(char* argv[]){
	cout << "retrieveDataFileReader()" <<endl;
	string input_path = argv[1];
	string input_filename = argv[2];
	string filename = input_path + "/" + input_filename;
	cout << "Creating Binary Dictionary For: " << filename <<endl;
	cout << "Final filename: " << filename << endl;
	return filename;
}

map<int, std::map<string, std::vector<int> > > createTimeFrameDict(char* argv[], int &stride, std::vector<string> &TrajectoryPaths, string &TopologyPath, int &totalFrames){
	cout << "createTimeFrameDict()" <<endl;
	int numLinesRead = 0;
	std::map<int, std::map<string, std::vector<int> > > timelapse_interaction_by_chain_dict;
	string filename = retrieveDataFileReader(argv);
	ifstream f; 
	f.open(filename.c_str());
	if(!f){
		cout << "Interaction Results Not Found" << endl;
	}else{
		int currFrame = 0;
		string currProtein = "";
		std::string line;
		while(std::getline(f, line)){
			numLinesRead ++;
			// if(numLinesRead >1000000){
			// 	break;
			// } 
			if(numLinesRead %10000 == 0){
				cout << "Line: " << numLinesRead << " ** " << line << endl;
			}
			if(line.find("Stride:") != std::string::npos){
				stride = stringToInt(split(line, ':')[1]);
			}else if(line.find("TrajectoryPath:") != std::string::npos){
				TrajectoryPaths.push_back(split(line, ':')[1]);
			}else if(line.find("TopologyPath:") != std::string::npos){
				TopologyPath = split(line, ':')[1];
			}else if(line.find("nFrames:") != std::string::npos){
				continue;
			}else if(line.find("Computing Time:") != std::string::npos){
				continue;
			}else if(line.find("Frame:") != std::string::npos){
				currFrame = stringToInt(split(line, ':')[1]);
			}else if(line.find("Salt Bridges:") != std::string::npos){
				currProtein = split(line, ':')[1];
			}else if(line.find("Pi-Cation:") != std::string::npos){
				currProtein = split(line, ':')[1];
			}else if(line.find("Pi-Stacking:") != std::string::npos){
				currProtein = split(line, ':')[1];
			}else if(line.find("T-Stacking:") != std::string::npos){
				currProtein = split(line, ':')[1];
			}else if(line.find("Van Der Waals:") != std::string::npos){
				currProtein = split(line, ':')[1];
			}else if(line.find("Hydrogen Bonds:") != std::string::npos){
				currProtein = split(line, ':')[1];
			}else if(line.find("Hydrogen Bond-Water Mediated:") != std::string::npos){
				currProtein = split(line, ':')[1];
			}else if(line.find("Backbone Backbone Hydrogen Bonds:") != std::string::npos){
				currProtein = split(line, ':')[1];
			}else if(line.find("Sidechain Backbond Hydrogen Bonds:") != std::string::npos){
				currProtein = split(line, ':')[1];
			}else if(line.find("Sidechain Sidechain Hydrogen Bonds:") != std::string::npos){
				currProtein = split(line, ':')[1];
			}else if(line.find("Residue Water Hydrogen Bonds:") != std::string::npos){
				currProtein = split(line, ':')[1];
			}else if(line.find("Water Bonds:") != std::string::npos){
				currProtein = split(line, ':')[1];
			}else if(line.find("Extended Water Bonds:") != std::string::npos){
				currProtein = split(line, ':')[1];
			}
			// Ligand Based Interactions
			else if(line.find("Ligand Backbone Hydrogen Bonds:") != std::string::npos){
				currProtein = split(line, ':')[1];
			}else if(line.find("Ligand Sidechain Hydrogen Bonds:") != std::string::npos){
				currProtein = split(line, ':')[1];
			}else if(line.find("Ligand Water Hydrogen Bonds:") != std::string::npos){
				currProtein = split(line, ':')[1];
			}else if(line.find("Ligand Water Bonds:") != std::string::npos){
				currProtein = split(line, ':')[1];
			}else if(line.find("Ligand Extended Water Bonds:") != std::string::npos){
				currProtein = split(line, ':')[1];
			}
			else if(!line.empty()){
				string key = line;
				int chain_id = getChainId(line);
				if(timelapse_interaction_by_chain_dict.find(chain_id) == timelapse_interaction_by_chain_dict.end()){
					std::map<string, std::vector<int> > internalDict; //interaction keys to value = [1,0,0,0,1...]
					std::vector<int> presentTimeFrames;
					presentTimeFrames.push_back(currFrame);
					internalDict.insert(std::pair<string, std::vector<int> >(key, presentTimeFrames));
					timelapse_interaction_by_chain_dict.insert(std::pair<int, std::map<string, std::vector<int> > >(chain_id, internalDict));
				}else{
					if(timelapse_interaction_by_chain_dict[chain_id].find(key) == timelapse_interaction_by_chain_dict[chain_id].end()){
						std::vector<int> presentTimeFrames;
						presentTimeFrames.push_back(currFrame);
						timelapse_interaction_by_chain_dict[chain_id].insert(std::pair<string, std::vector<int > >(key, presentTimeFrames));
					}else{
						timelapse_interaction_by_chain_dict[chain_id][key].push_back(currFrame);
					}
				}
			}
		}
		totalFrames = currFrame + 1;
	}
	return timelapse_interaction_by_chain_dict;
}

string writeRootPath(string stitchedInteractionPath){
	size_t len = strlen("mOR-InteractionOutput");
	size_t found = stitchedInteractionPath.find("mOR-InteractionOutput");
	if(found != std::string::npos){
		stitchedInteractionPath.replace(found, len, "mOR-InteractionOutputDictionary");
	}
	return stitchedInteractionPath;
}


void writeFrequencyMap(char*argv[], int stride, std::vector<string> &TrajectoryPaths, string &TopologyPath, 
						std::map<int,std::map<string, std::vector<int> > > & timelapse_interaction_by_chain_dict, int &totalFrames){
	cout << "writeFrequencyMap()" << endl;
	int numKeysWritten = 0;
	string output_path = argv[3];
	string output_filename = argv[4];
	string interaction_selection = argv[5];
	string filename = output_path + "/" + output_filename;
	string mkcmd = "mkdir " + output_path;
	system(mkcmd.c_str());
	string buffer;
	ofstream f;
	f.open(filename.c_str());
	f << "Stride:" << stride << endl;
	for(int i = 0; i<TrajectoryPaths.size(); i++){
		string TrajectoryPath = TrajectoryPaths[i];
		f <<"TrajectoryPath:" << TrajectoryPath <<endl;
	}
	f << "TopologyPath:" << TopologyPath << endl;
	f << FLAG_INTERACTION[interaction_selection] << " Dictionary Heat Map " <<endl;
	f << "TotalFrames: " << totalFrames << endl;
	for(std::map<int, std::map<string, std::vector<int> > >::iterator it1 = timelapse_interaction_by_chain_dict.begin(); it1 != timelapse_interaction_by_chain_dict.end(); ++it1){
		int chain_id = it1->first;
		f << "Dictionary for Chain: " << chain_id << endl;
		std::map<string, std::vector<int> > timeFrameDict = it1->second;
		for(std::map<string, std::vector<int> >::iterator it2 = timeFrameDict.begin(); it2 != timeFrameDict.end(); ++it2){
			if(numKeysWritten % 1000 == 0){
				cout <<"Key #: " << numKeysWritten << endl;
			}
			string interaction_key = it2->first;
			f << interaction_key << "~[";
			std::vector<int> bitVector = it2->second;
			int bitVecSize = bitVector.size();
			for (int i = 0; i<bitVecSize; i++){
				if(i < bitVecSize-1){
					f << bitVector[i] << ",";
				}
			}
			f<<bitVector[bitVecSize-1] << "]" <<endl;
			numKeysWritten++;
		}
	}
	cout << "Wrote results to: " << filename << endl;
}

//Should probably write directly to file rather than saving in intermediary map. Memory storage too high 
std::vector<int> binarizeTime(std::vector<int> timeFrames, int totalFrames){
	vector<int> binaryVec;
	for (int i = 0; i<totalFrames; i++){
		binaryVec.push_back(0);
	}
	for (int j = 0; j<timeFrames.size(); j++){
		int frameIndex = timeFrames[j];
		binaryVec[frameIndex] = 1;
	}
	return binaryVec;
}


void binarizeAndWriteTimePoints(char*argv[], int stride, std::vector<string> &TrajectoryPaths, string &TopologyPath, 
																				std::map<int,std::map<string, std::vector<int> > > timelapse_interaction_by_chain_dict, 
																				int totalFrames){
	cout << "binarizeTimePoints()" << endl;
	int numKeysWritten = 0;
	string output_path = argv[3];
	string output_filename = argv[4];
	string interaction_selection = argv[5];
	string filename = output_path + "/" + output_filename;
	string buffer;
	ofstream f;
	f.open(filename.c_str());
	f << "Stride:" << stride << endl;
	for(int i = 0; i<TrajectoryPaths.size(); i++){
		string TrajectoryPath = TrajectoryPaths[i];
		f <<"TrajectoryPath:" << TrajectoryPath <<endl;
	}
	f << "TopologyPath:" << TopologyPath << endl;
	f << FLAG_INTERACTION[interaction_selection] << " Dictionary Heat Map " <<endl;
	f << "TotalFrames: " << totalFrames << endl;
	for(std::map<int, std::map<string, std::vector<int> > >::iterator it1 = timelapse_interaction_by_chain_dict.begin(); it1 != timelapse_interaction_by_chain_dict.end(); ++it1){
		int chain_id = it1->first;
		f << "Dictionary for Chain: " << chain_id << endl;
		std::map<string, std::vector<int> > timeFrameDict = it1->second;
		for(std::map<string, std::vector<int> >::iterator it2 = timeFrameDict.begin(); it2 != timeFrameDict.end(); ++it2){
			if(numKeysWritten % 1000 == 0){
				cout <<"Key #: " << numKeysWritten << endl;
			}
			string interaction_key = it2->first;
			std::vector<int> timeFrames = it2->second;
			std::vector<int> binaryVec = binarizeTime(timeFrames, totalFrames);
			f << interaction_key << "~[";
			int bitVecSize = binaryVec.size();
			for(int i = 0; i<bitVecSize;i++){
				if(i<bitVecSize-1){
					f << binaryVec[i] << ",";
				}
			}
			f<<binaryVec[bitVecSize-1] <<"]" <<endl;
			numKeysWritten ++;
		}
	}
	cout << "Wrote results to: " << filename << endl;
}


void initialize_flags(){
	FLAG_INTERACTION.insert(std::pair<string, string> ("-sb", "Salt Bridges"));
	FLAG_INTERACTION.insert(std::pair<string, string> ("-pc", "Pi-Cation"));
	FLAG_INTERACTION.insert(std::pair<string, string> ("-ps", "Pi-Stacking"));
	FLAG_INTERACTION.insert(std::pair<string, string> ("-ts", "T_Stacking"));
	FLAG_INTERACTION.insert(std::pair<string, string> ("-vdw", "Van Der Waals"));
	FLAG_INTERACTION.insert(std::pair<string, string> ("-hb", "Hydrogen Bonds"));
	FLAG_INTERACTION.insert(std::pair<string, string> ("-hbw", "Hydrogen Bond-Water Mediated"));
	FLAG_INTERACTION.insert(std::pair<string, string> ("-hbbb", "Backbone Backbone Hydrogen Bonds"));
	FLAG_INTERACTION.insert(std::pair<string, string> ("-hbsb", "Sidechain Backbond Hydrogen Bonds"));
	FLAG_INTERACTION.insert(std::pair<string, string> ("-hbss", "Sidechain Sidechain Hydrogen Bonds"));
	FLAG_INTERACTION.insert(std::pair<string, string> ("-rw", "Residue Water Hydrogen Bonds"));
	FLAG_INTERACTION.insert(std::pair<string, string> ("-wb", "Water Bonds"));
	FLAG_INTERACTION.insert(std::pair<string, string> ("-wb2", "Extended Water Bonds"));
	// Ligand Based Interactions
	FLAG_INTERACTION.insert(std::pair<string, string> ("-hlb", "Ligand Backbone Hydrogen Bonds"));
	FLAG_INTERACTION.insert(std::pair<string, string> ("-hls", "Ligand Sidechain Hydrogen Bonds"));
	FLAG_INTERACTION.insert(std::pair<string, string> ("-lw", "Ligand Water Hydrogen Bonds"));
	FLAG_INTERACTION.insert(std::pair<string, string> ("-lwb", "Ligand Water Bonds"));
	FLAG_INTERACTION.insert(std::pair<string, string> ("-lwb2", "Ligand Extended Water Bonds"));

}

int main(int argc, char* argv[]){
	initialize_flags();
	int stride = 1;
	std::vector<string> TrajectoryPaths;
	string TopologyPath = "";
	int totalFrames = 0;
	std::map<int,std::map<string, std::vector<int> > > timelapse_interaction_by_chain_dict =  createTimeFrameDict(argv, stride, TrajectoryPaths, TopologyPath, totalFrames);
	cout<< "Number of Distinct Keys: " << timelapse_interaction_by_chain_dict[0].size() << endl;
	if(argc > 6){
		string binarizeFlag = argv[6];
		if(binarizeFlag.find("-b") != std::string::npos){
			binarizeAndWriteTimePoints(argv, stride, TrajectoryPaths, TopologyPath, timelapse_interaction_by_chain_dict, totalFrames);
		}
	}else{
		writeFrequencyMap(argv, stride, TrajectoryPaths, TopologyPath, timelapse_interaction_by_chain_dict, totalFrames);
	}
	return 0;
}

